src/Servers/Kestrel/Transport.MsQuic/Internal/QuicStream.cs

```diff
@@ -0,0 +1,187 @@
+
+using System;
+using System.Buffers;
+using System.Diagnostics;
+using System.Runtime.InteropServices;
+using System.Threading.Tasks;
+using static Microsoft.AspNetCore.Server.Kestrel.Transport.MsQuic.Internal.NativeMethods;
+
+namespace Microsoft.AspNetCore.Server.Kestrel.Transport.MsQuic.Internal
+{
+    internal sealed class QuicStream : IDisposable
+    {
+        private bool _disposed;
+        private readonly bool _shouldOwnNativeObj;
+        private IntPtr _nativeObjPtr;
+        private GCHandle _handle;
+        private StreamCallback _callback;
+        private readonly IntPtr _unmanagedFnPtrForNativeCallback;
+
+        public delegate QUIC_STATUS StreamCallback(
+            QuicStream stream,
+            ref StreamEvent evt);
+
+        internal StreamCallbackDelegate NativeCallback { get; private set; }
+
+        internal static QUIC_STATUS NativeCallbackHandler(
+           IntPtr stream,
+           IntPtr context,
+           ref StreamEvent connectionEventStruct)
+        {
+            var handle = GCHandle.FromIntPtr(context);
+            var quicStream = (QuicStream)handle.Target;
+
+            return quicStream.ExecuteCallback(ref connectionEventStruct);
+        }
+
+        private QUIC_STATUS ExecuteCallback(
+            ref StreamEvent evt)
+        {
+            var status = QUIC_STATUS.INTERNAL_ERROR;
+            if (evt.Type == QUIC_STREAM_EVENT.SEND_COMPLETE)
+            {
+                SendContext.HandleNativeCallback(evt.ClientContext);
+            }
+            try
+            {
+                status = _callback(
+                    this,
+                    ref evt);
+            }
+            catch (Exception)
+            {
+                // TODO log
+            }
+            return status;
+        }
+
+        public void SetCallbackHandler(
+            StreamCallback callback)
+        {
+            _handle = GCHandle.Alloc(this);
+            _callback = callback;
+            Registration.SetCallbackHandlerDelegate(
+                _nativeObjPtr,
+                _unmanagedFnPtrForNativeCallback,
+                GCHandle.ToIntPtr(_handle));
+        }
+
+        public unsafe Task SendAsync(
+           ReadOnlySequence<byte> buffers,
+           QUIC_SEND_FLAG flags,
+           object clientSendContext)
+        {
+            var sendContext = new SendContext(buffers, clientSendContext);
+            var status = (QUIC_STATUS)Registration.StreamSendDelegate(
+                _nativeObjPtr,
+                sendContext.Buffers,
+                sendContext.BufferCount,
+                (uint)flags,
+                sendContext.NativeClientSendContext);
+            QuicStatusException.ThrowIfFailed(status);
+
+            // TODO for some reason the first call to SendAsync triggers the callback.
+            return sendContext.TcsTask; // TODO make QuicStream implement IValueTaskSource?
+        }
+
+        public Task StartAsync(QUIC_STREAM_START flags)
```

Comment:
Start is only async if you pass the async flag.

Reply:
Which flag are you referring to?

Reply:
`QUIC_STREAM_START_FLAG_ASYNC`

Reply:
Oh yeah, that was a new one. I'll do the needful.